aggregator/aggregator.go

```diff
@@ -46,395 +60,776 @@ func NewAggregator(
 		profitabilityChecker = NewTxProfitabilityCheckerAcceptAll(stateInterface, cfg.IntervalAfterWhichBatchConsolidateAnyway.Duration)
 	}
 
-	proverClients := make([]proverClientInterface, 0, len(cfg.ProverURIs))
-	ctx := context.Background()
-
 	a := Aggregator{
 		cfg: cfg,
 
-		State:                stateInterface,
-		EthTxManager:         ethTxManager,
-		Ethman:               etherman,
-		ProverClients:        proverClients,
-		ProfitabilityChecker: profitabilityChecker,
+		State:                   stateInterface,
+		EthTxManager:            ethTxManager,
+		Ethman:                  etherman,
+		ProfitabilityChecker:    profitabilityChecker,
+		StateDBMutex:            &sync.Mutex{},
+		TimeSendFinalProofMutex: &sync.RWMutex{},
+
+		finalProof: make(chan finalProofMsg),
 	}
 
-	for _, proverURI := range cfg.ProverURIs {
-		proverClient := prover.NewClient(proverURI, cfg.IntervalFrequencyToGetProofGenerationState)
-		proverClients = append(proverClients, proverClient)
-		grpcClientConns = append(grpcClientConns, proverClient.Prover.Conn)
-		log.Infof("Connected to prover %v", proverURI)
+	return a, nil
+}
 
-		// Check if prover is already working in a proof generation
-		proof, err := stateInterface.GetWIPProofByProver(ctx, proverURI, nil)
-		if err != nil && err != state.ErrNotFound {
-			log.Errorf("Error while getting WIP proof for prover %v", proverURI)
-			continue
-		}
+// Start starts the aggregator
+func (a *Aggregator) Start(ctx context.Context) error {
+	var cancel context.CancelFunc
+	if ctx == nil {
+		ctx = context.Background()
+	}
+	ctx, cancel = context.WithCancel(ctx)
+	a.ctx = ctx
+	a.exit = cancel
 
-		if proof != nil {
-			log.Infof("Resuming WIP proof generation for batchNumber %v in prover %v", proof.BatchNumber, *proof.Prover)
-			go func() {
-				a.resumeWIPProofGeneration(ctx, proof, proverClient)
-			}()
-		}
+	// Delete ungenerated recursive proofs
+	err := a.State.DeleteUngeneratedProofs(ctx, nil)
+	if err != nil {
+		return fmt.Errorf("Failed to initialize proofs cache %w", err)
 	}
 
-	a.ProverClients = proverClients
+	address := fmt.Sprintf("%s:%d", a.cfg.Host, a.cfg.Port)
+	lis, err := net.Listen("tcp", address)
+	if err != nil {
+		log.Fatalf("failed to listen: %v", err)
+	}
 
-	return a, nil
+	a.srv = grpc.NewServer()
+	pb.RegisterAggregatorServiceServer(a.srv, a)
+
+	healthService := newHealthChecker()
+	grpc_health_v1.RegisterHealthServer(a.srv, healthService)
```

Comment:
This naming does not look fine to me. I would remove the v1 from it everywhere.

Reply:
@ToniRamirezM I agree that they choose an ugly package name :smile: 
I added an alias here dbdc312e7eb3f0a7c9a8d0e1088102cbe469fc5f, does it sound better to you?